File: week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.py
# Uses python3
import sys

def select_max_index(values,weights):
    ma=-1
    j=-1
    for i in range(n):
        if weights[i]!=0:
            if ma<values[i]/weights[i]:
                ma=values[i]/weights[i]
                j=i
    return j

def get_optimal_value(capacity, weights, values):
    value = 0.

    if capacity == 0:
        return 0
    for i in range(n):
        max_index = select_max_index(values, weights)
        if max_index >= 0:
            available_weights = min(capacity, weights[max_index])
            value = value + available_weights * values[max_index]/weights[max_index]
            weights[max_index] = weights[max_index] - available_weights
            capacity = capacity - available_weights
            
    return value


if __name__ == "__main__":
    data = list(map(int, sys.stdin.read().split()))
    n, capacity = data[0:2]
    values = data[2:(2 * n + 2):2]
    weights = data[3:(2 * n + 2):2]
    opt_value = get_optimal_value(capacity, weights, values)
    print("{:.10f}".format(opt_value))

File: week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.cpp
#include <iostream>
#include <vector>

using std::vector;

double get_optimal_value(double capacity, vector<double> weights, vector<double> values, int n)
{
  double value = 0.0;
  while(capacity && n)
  {
      double maxi=0;
      int j=-1;
      for(int i=0;i<n;i++)
      {
          //if(weights[i]!=0)
            if(maxi<(values[i]/weights[i]))
            {
                maxi=values[i]/weights[i];
                j=i;
            }
      }

      if(capacity<weights[j])
      {
          value+=capacity*values[j]/weights[j];
          capacity-=capacity;
      }
      else
      {
          value+=values[j];
          capacity-=weights[j];
          //weights[j]=0;
      }

      n--;
      for(int i=j;i<n;i++)
      {
          weights[i]=weights[i+1];
          values[i]=values[i+1];
      }
  }
  return value;
}

int main()
{
  int n;
  double capacity;
  std::cin >> n >> capacity;
  vector<double> values(n);
  vector<double> weights(n);
  for (int i = 0; i < n; i++)
    std::cin >> values[i] >> weights[i];

  double optimal_value = get_optimal_value(capacity, weights, values, n);
  std::cout.precision(10);
  std::cout << optimal_value << std::endl;
  return 0;
}
